File: Practica 1/src/gestorAplicacion/Persona/Cliente.java
package gestorAplicacion.Persona;
import java.io.Serializable;
import java.util.*;
import gestorAplicacion.Taquilla.*;
import gestorAplicacion.Autocine.*;

/**
 * @author Jimena Uribe Giraldo.
 * @summary Clase Cliente, lleva todo lo relativo a la información del espectador
 *
 */
public class Cliente implements Serializable {
 
	// Serializacion.
	 	private static final long serialVersionUID = 1L;
	 	static List<Cliente> clientess;
	 	static {
	 		clientess = new ArrayList<Cliente>();
	 	}
	 	
    /*Atributos*/
    private int id;
    private String nombre;
    private static int edad;
    private static List<Ticket> historialCompras= new ArrayList<Ticket>(); //una lista con los boletos que ha comprado el cliente en su vida
    private Autocine autocine;
    
 
	/*Constructores*/
    public Cliente(int id, String nombre, int edad, Autocine autocine) {
		this.id = id;
		this.nombre = nombre;
		Cliente.edad = edad;
		this.autocine=autocine;
	}
    
    public String GeneroMasVisto() {
		/*
		No recibe nada y devuelve string del género más visto del cliente
		 */
		List<String> genreList=new ArrayList<String>();		//lista con los generos que ha visto el cliente
		for(Ticket ticket: historialCompras) {
			Ticket.getFuncion();
			genreList.add(Funcion.getPelicula().getGenero()); 	//Recorre el historialdel cliente y anexa sus generos
		}
		List<Integer> veces=new ArrayList<Integer>();		//lista para guardar la frecuencia de cada genero
		for(String genre: genreList) {
			int occ = Collections.frequency(genreList, genre);	//De la lista de géneros extrae la frecuencia
			veces.add(occ);
		}
		
		return genreList.get(veces.indexOf(Collections.max(veces))); 	//devuelve el género más visto
		
	}
    
 
    /*set y get*/
    public String getNombre() {
        return nombre;
    }
 
    public void setNombre(String nombre) {
        this.nombre = nombre;
    }
 
    public static int getEdad() {
        return edad;
    }
 
    public void setEdad(int edad) {
        Cliente.edad = edad;
    }
 
    public int getId() {
        return id;
    }
 
    public void setId(int id) {
        this.id = id;
    }
 
    
    public static List<Ticket> getHistorialCompras() {
		return historialCompras;
	}
	public void setHistorialCompras(List<Ticket> historialCompras) {
		Cliente.historialCompras = historialCompras;
	}


	public Autocine getAutocine() {
		return autocine;
	}

	public void setAutocine(Autocine autocine) {
		this.autocine = autocine;
	}
	
	public static List<Cliente> getClientes() {
		return clientess;
	}

    
    @Override
	public String toString() {
		
		return "Cliente: " + nombre + "-" + String.valueOf(edad);
		
	}
 
}

File: Practica 2/AutoCine/src/gestorAplicacion/Persona/Cliente.py
from gestorAplicacion.Taquilla import *
from gestorAplicacion.Taquilla.Ticket import Ticket
from gestorAplicacion.Autocine.Autocine import Autocine
from gestorAplicacion.Persona import *

from typing import Collection
from collections import Counter

#*
# * @author Jimena Uribe Giraldo.
# * @summary Clase Cliente, lleva todo lo relativo a la información del espectador
# 
# 



class Cliente:
    
    def __init__(self,id,nombre,edad,autocine):
        self._historialCompras = [] #una lista con los boletos que ha comprado el cliente en el cine a traves de su vida
        self._id= id
        self._nombre = nombre
        self._edad = edad
        Autocine.agregarCliente(self)
        self._autocine = autocine

    def GeneroMasVisto(self):
        #No recibe nada y devuelve string del género más visto del cliente

        genreList = [] #lista con los generos que ha visto el cliente
        for ticket in self._historialCompras:
            genreList.append(Ticket.getFuncion().getPelicula().getGenero()) #Recorre el historial de compras del cliente y anexa los generos de los tickets

        veces=Counter(genreList).items()   #lista para guardar la frecuencia de cada genero

        occ=Counter(genreList).values() #De la lista de géneros extrae la frecuencia de cada elemento
        valor_max=max(occ)
        
        for genero  in veces:
            if genero[1]==valor_max:
                return genero[0]



    #
    #Getting and setting
    #
    def getId(self):
        return self._id
    def setId(self, id):
        self._cedula = id

    def getNombre(self):
        return self._nombre
    def setNombre(self, nombre):
        self._nombre = nombre

    def getEdad(self):
        return self._edad
    def setEdad(self, edad):
        self._edad = edad

  

    def getHistorialCompras(self):
        return self._historialCompras
    def setHistorialCompras(self, historialCompras):
        self._historialCompras = historialCompras


    def getAutocine(self):
        return self._autocine
    def setAutoine(self, autocine):
        self._autocine = autocine
    
